Fix performance issue with UTF8ToString (#12517)

LanderlYoung opened the pull request:
~- add extra parameter `exactStringLength` to UTF8ToString,~
~This parameter is optional, and if given, the `maxBytesToRead` parameter is ignored.~
~to keep consistent API flavor, `UTF16ToString`, `UTF32ToString`, is also changed.~

    Fix performance issue with UTF8ToString (#12517)
    
    introduce new functions
    - UTF8ToStringWithLength
    - UTF16ToStringWithLength
    - UTF32ToStringWithLength
    
    Decode string exactly length with given length, any '\0' in between will
    be kept as-is.
    
    Those functions require an argument `lengthInBytes`, so no need to iterator
    the heap to find a null-terminator, thus have better performance.

juj commented:
This regresses generated code size for users that do not use the new variant.

Also unless I am mistaken, this does not really fix any performance issue, because no caller code site is actually changed to use the exact length form. The PR just slows down the functions when exact length is not specified.

Maybe this fixes a performance issue in your JS library code calls that are updated to use this form, at the expense of slowing down generated code for every other user?

I would recommend implementing a separate set of functions `UTF*ToStringExactNBytes` where `maxBytesToRead` would be replaced with a non-optional parameter `exactBytesToRead`. Then callers can use that variant when they need to marshal exactly requested number of bytes.

For the new variant, add a clear comment that if there are null bytes in the middle of the string, then those null bytes will also be deserialized into the generated string.

LanderlYoung commented:
@juj Thanks for the suggestion. BTW: previous discussion here https://github.com/emscripten-core/emscripten/issues/12517

This does not fix the performance issue, because the function still needs to be backward compatible.
But it provides a way to be more efficient.

> ...at the expense of slowing down generated code for every other user?

I don't think this PR would slow down other's code. Just added some simple if checks.


> I would recommend implementing a separate set of functions

This is nice.

juj commented:
> I don't think this PR would slow down other's code. Just added some simple if checks.

The loop condition changed from

```
while (idx < endPtr)
```
to
```
    while (hasExactLength ? idx < endPtr : !(idx >= endIdx)) {
```
which will slow down code for all users.

Let's go with a separate set of functions, that will be best for code size and performance in each case.

LanderlYoung commented:
Ok, new commit latter

LanderlYoung commented:
@juj I rewrote this PR and added corresponding UnitTest, please review.

LanderlYoung commented:
BTW, the  tests/utf32.cpp file didn't properly formated with clang-format, and since this file is small, and my test now makes up more lines, I formatted it with project's .clang-format config.

LanderlYoung commented:
Some Circle CI builds failed, but as I tested it also fails on master (like `test_relocatable_void_function`).

Help needed.

juj commented on src/runtime_strings.js:
```diff
@@ -118,6 +118,85 @@ function UTF8ToString(ptr, maxBytesToRead) {
+  return ptr ? UTF8ArrayToStringWithLength(HEAPU8, ptr, lengthInBytes) : '';
+#endif
+}
+
```
Hmm, the name `StringWithLength` parses to my eye as it would be the "JavaScript string with length", which is odd. That is, the function should more exactly be `UTF8WithLengthToString` if the word `with` is used?

Would either one of names `UTF8ToStringNBytes`, `UTF8ToStringExactNBytes`, `UTF8ToStringNB` be better to avoid making the connotation directly to string, but rather to the operation overall?

LanderlYoung replied:
Naming is hard！I'll think about it.

Found a typo in 'as-it' , should be 'as-is', will fix it.

LanderlYoung replied:
For now I may personally prefer `UTF8ToStringNBytes`. 😛

sbc100 replied:
`UTF8ToStringN` or `UTF8ToStringNBytes` both sounds reasonable.

What happens if there is a null bytes in the first N bytes of the string?   Can JS strings contains nulls?

LanderlYoung replied:
Yes, JS String can contain null. Here is a simple test code:

```js
const decoder = new TextDecoder("utf-8");

const raw = new Int8Array([
    0,  // '\0'
    0,  // '\0'
    0,  // '\0'
    65, // 'A'
    65, // 'A'
    0,  // '\0'
    65, // 'A'
    0,  // '\0'
    0,  // '\0'
]);

const str = "\0\0\0AA\0A\0\0";

const decodedString = decoder.decode(raw);

console.assert(str[0] === '\0');
console.assert(str[3] === 'A');
console.assert(str[7] === '\0');
console.assert(str === decodedString);

```

juj commented on src/runtime_strings.js:
```diff
@@ -118,6 +118,85 @@ function UTF8ToString(ptr, maxBytesToRead) {
 #endif
 }
 
+
```
Please move all the added content in this file over to `src/library_strings.js`. The file `runtime_strings.js` should not receive any more content, it exists for historical reasons. All new functions can go to `library_strings.js`.

LanderlYoung replied:
accepted

LanderlYoung replied:
I tried to move code to `src/library_strings.js`, but it seems to have something to be related to the `MINIMAL_RUNTIME` flag, no matter how the unit test complains `ReferenceError: UTF8ToStringNBytes is not defined`.

The half-done commit is here https://github.com/LanderlYoung/emscripten/commit/f9149d794f38beedb8d4d6ff6c0d53ebb6cfbec6

If there is and doc related, please let me know, thanks.

sbc100 replied:
The way the `library_xxx.js` system works is that symbols are only ever included on on demand.

There are several ways a library symbol can get included:

1. By being directly referenced by native code
2. By being add to `-s  DEFAULT_LIBRARY_FUNCS_TO_INCLUDE` list.
3. Via the `__deps` entry to another symbol that is incuded.

(1) doesn't apply to JS functions that are only callable from JS such as these ones so you can ignore that.  To signal these are JS-only functions you should prefix them with `$`.

sbc100 replied:
It looks like `src/library_strings.js` is only currently included in MINIMAL_RUNTIME mode.  Its job is currently to make the functions in `runtime_strings_extra.js` available as JS library, but only MINIMAL_RUNTIME mode.

`runtime_strings_extra.js` is include directly in normal (non-MINIMAL) mode.

So some refactoring would need to be doing before these functions can be make universally available via the library system.

I'd be tempted to add all these new functions to `runtime_strings_extra.js` + `src/library_strings.js` for now.  At least then they will be used in library form under MINIMAL_RUNTIME.

The alternative is that we do some refactoring such at `src/library_strings.js` is used in all modes.

juj commented on src/runtime_strings_extra.js:
```diff
@@ -77,6 +77,37 @@ function UTF16ToString(ptr, maxBytesToRead) {
 #endif // TEXTDECODER
 }
 
+function UTF16ToStringWithLength(ptr, lengthInBytes) {
```
Please move all the added code from this file into file `src/library_strings.js`. This file exists only for historical compatibility reasons, but all new code can be added to `src/library_strings.js`.

LanderlYoung replied:
accepted

juj commented:
Great work! Thanks for championing this. One code organization comment only, we don't need to use the `runtime_x.js` files for new code, but can use the proper `library_x.js` structuring there.

LanderlYoung commented:
update:

Fixed cr issues:

1. renamed `UTF*ToStringWithLength` to `UTF*ToStringNBytes`
2. fixed typo in doc
3. add docs to `preamble.js.rst`

BUT, failed to move code to `library_strings.js`, more detail explained in [cr conversation](https://github.com/emscripten-core/emscripten/pull/12544#discussion_r509222535).
